### ISSUE TITLE
Switch to ajv/2020 for fastify 5

### ISSUE THREAD

**Issue by kibertoad** [fastify/fastify#5448]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

I wonder if we should consider switching to `ajv/dist/2020` during our next SemVer major.

According to https://github.com/sinclairzx81/typebox/issues/490#issuecomment-1621256750, default ajv is still using Draft 7 version of JSON schema.

### Motivation

Support latest additions to the JSON schema standard. Considering the scope and impact of this change, I doubt we can do this outside of a SemVer major.

### Example

I don't expect usage patterns for ajv or ajv-compiler to change, it will simply allow to support the features introduced in later versions of JSON Schema.

**Comment by climba03003** [comment]
Draft 7 and Draft 2020-12 is great different.
It also provides a large set of incompatibility.

**Comment by kibertoad** [comment]
@climba03003 Do you mean that existing Draft 7 schemas will stop working under Draft 2020?

**Comment by climba03003** [comment]
> Do you mean that existing Draft 7 schemas will stop working under Draft 2020?

Mostly, yes.

https://json-schema.org/draft/2019-09/release-notes
https://json-schema.org/draft/2020-12/release-notes

**Comment by metcoder95** [comment]
Draft-2020 is also marked as `non-backwards compatible` by `ajv`: https://ajv.js.org/json-schema.html#draft-2020-12

It would be nice but a lot of schemas uses mostly the 2019-draft, and doing this make them prone to break and cause friction with the adoption of Fastify v5.

Might be better to allow them do it if they want or still use 2019's draft

**Comment by gurgunday** [comment]
It's also important to note that, and this is what changed my mind on this, the AJV docs recommend sticking with 07 if we don't need the two new properties as **all schemas will be affected by the reduced performance required to support them**

https://github.com/fastify/ajv-compiler/issues/105#issuecomment-1722434616

**Comment by Eomm** [comment]
Switching the ajv version is as easy as writing:

```
const factoryValidator = require('@fastify/ajv-compiler')()

const app = fastify({
  schemaController: {
    compilersFactory: {
      buildValidator: factoryValidator
    }
  }
})
```

Since changing the default would be a massive breaking change for the user (as it should rewrite all its schemas) - we could just ship a new `@fastify/ajv-compiler` major with `ajv/2020` and let those users that want the new standard set the `buildValidator` option accordingly.

**Comment by mcollina** [comment]
I think we would need fast-json-stringify to support it first.

**Comment by imjuni** [comment]
When upgrading the version of ajv, I often encountered difficulties due to [@fastify/swagger](https://github.com/fastify/fastify-swagger). @fastify/swagger is an important third-party plugin but only supports JSON Schema draft-07. While @fastify/swagger is flexible with JSON Schema versions, [json-schema-resolver](https://www.npmjs.com/package/json-schema-resolver) only supports [draft-07](https://github.com/Eomm/json-schema-resolver/blob/63020a0dbaaac1b088e16e6e17435af8670cd8af/ref-resolver.js#L26).

**Comment by jviotti** [comment]
Hey there! JSON Schema TSC here, and I accidentally came across this thread. 

To address the compatibility questions, good JSON Schema implementations support multiple JSON Schema dialects at the same time. Depending on your `$schema` value, the same JSON Schema implementation should be able to work with any schema given to it.

In any case, keep in mind AJV is known to be very non-compliant, in fact being one of the worst. See https://bowtie.report/#/implementations/js-ajv for its official compliance scores across tested dialects. Right now, AJV is failing on 661 compliance test cases (quite a lot). The problem though is that there are not many good JSON Schema implementations that are fast. AJV gets away with its lack of compliance because it is pretty fast.

I'm personally working on a high performance validator capable of nano-second schema validation (>200x faster than AJV) that has 100% compliance scores and supports every dialect of JSON Schema at once. It's C++ and mainly targeted at low-latency API gateways, but I'm experimenting with doing JS bindings without sacrificing on performance. Would be fun to try a Fastify plugin at some point!

**Comment by climba03003** [comment]
> Right now, AJV is failing on 661 compliance test cases (quite a lot).

When I see this test to be failed, 
![image](https://github.com/user-attachments/assets/b9f39d9e-8f4d-41ce-9e75-db919cc3514d)

it seems to be unfair on `ajv` since the format part is supported by plugin.
![image](https://github.com/user-attachments/assets/3677135b-33a6-49f4-a910-1708ad2c6690)

**Comment by jviotti** [comment]
@climba03003 That error is different. Is testing that if you set `format` (whether supported or not) but you pass something that is not a string, it must succeed, unless `type: string` is also present. Even without the plugin, that is non-compliant behaviour.

**Comment by gurgunday** [comment]
> In case its interesting to anybody, I'm personally working on a high performance validator capable of nano-second schema validation (>200x faster than AJV) that has 100% compliance scores and supports every dialect of JSON Schema at once. It's C++ and mainly targeted at low-latency API gateways, but I'm experimenting with doing JS bindings without sacrificing on performance. Would be fun to try a Fastify plugin at some point!

This sounds awesome, please do release a node version

I don't think we ship binaries in general, but it could indeed be interesting who knows

**Comment by jviotti** [comment]
We are experimenting with an MIT/Apache-2.0 distribution running on WebAssembly, without involving Node.js native add-ons. I'll share the link once we have it!

**Comment by Relequestual** [comment]
May wish to consider something which is more compliant by default.
I eagarly await @jviotti's validator in wasam. It may be a little expensive on application load. If Node.js bindings materialise, great!

Otherwise, might be worth looking into the very compliant https://github.com/hyperjump-io/json-schema
Will depend on what features from AJV you'd consider essential.

**Comment by jviotti** [comment]
:+1: on @Relequestual 's HyperJump suggestion. As an update on Blaze, we finished the last optimisations we wanted to do for its paper & launch a couple of weeks ago. We will be publishing the paper hopefully in the next month (showing Blaze significantly outperforming every other validator in every tested case) and will continue working on the bindings.

**Comment by gurgunday** [comment]
Yeah if I could find a proper benchmark for https://github.com/hyperjump-io/json-schema, I'd maybe propose it for Fastify v6, but I'm still really interested by Blaze
